### PR TITLE
Migrate concept fields to ontology definition

### PR DESCRIPTION
**Pull request by MiroslavGatsanoga**
# Description

## What

- Define all concept fields as ontology config and remove any unneeded code
- Add validation for the field values' type

## Why

[JIRA ticket](https://financialtimes.atlassian.net/browse/UPPSF-2617).

## Scope and particulars of this PR (Please tick all that apply)

- [ ] Tech hygiene (dependency updating & other tech debt)
- [ ] Bug fix
- [X] Feature
- [ ] Documentation
- [ ] Breaking change
- [ ] Minor change (e.g. fixing a typo, adding config)

___
This Pull Request follows the rules described in our [Pull Requests Guide](https://github.com/Financial-Times/upp-docs/tree/master/guides/pr-guide)

### CODE REVIEW

File: ontology/config.go [hunk 1]
```diff
@@ -15,9 +22,69 @@ type RelationshipConfig struct {
...
+		}
+
+		return true
+	case "int":
+		_, ok := val.(int)
+		if ok {
+			return true
+		}
+
+		_, ok = val.(float64) // float64, for JSON numbers
```
Comment (epavlova) [inline ontology/config.go:83]: **question:** Could this be misleading for the clients of the ontology? I understand why we are doing that, when unmarshalling numbers automatically are parsed as `float64`. But what happens if I'm a client of the ontology that uses the method and after everything goes well I decide to do unsafe type assertion to `int`. Just thinking out loud here, let's at least add some comments how the method is intended to be used. Or if you think of a better way to mitigate such scenarios?

File: ontology/config.yml [hunk 2]
```diff
@@ -1,4 +1,76 @@
-fieldToNeoProps:
+fields:
+  strapline:
+    neoProp: "strapline"
```
Comment (epavlova) [inline ontology/config.yml:3]: **suggestion:** What do you think about making `neoProp` mandatory only when the name used in Neo is different from the name of the field?

File: ontology/concept.go [hunk 3]
```diff
@@ -240,30 +203,12 @@ type NewConcept struct {
...
-	// Location
-	ISO31661 string `json:"iso31661,omitempty"`
-	// Person
-	Salutation string `json:"salutation,omitempty"`
-	BirthYear  int    `json:"birthYear,omitempty"`
-	// Industry Classifications
-	IndustryIdentifier string `json:"industryIdentifier,omitempty"`
 }
 
 var ErrEmptyAuthority = errors.New("invalid request, no sourceRepresentation.authority has been supplied")
```
Comment (epavlova) [inline ontology/concept.go:214]: **nitpick:** typo in `ErrUnkownAuthority` -> `ErrUnknownAuthority`

File: ontology/config.go [hunk 4]
```diff
@@ -15,9 +22,69 @@ type RelationshipConfig struct {
...
 type Config struct {
-	FieldToNeoProps map[string]string             `yaml:"fieldToNeoProps"`
-	Relationships   map[string]RelationshipConfig `yaml:"relationships"`
-	Authorities     []string                      `yaml:"authorities"`
+	Fields        map[string]FieldConfig        `yaml:"fields"`
+	Relationships map[string]RelationshipConfig `yaml:"relationships"`
+	Authorities   []string                      `yaml:"authorities"`
+}
+
+var ErrUnkownProperty = errors.New("unknown concept property")
```
Comment (epavlova) [inline ontology/config.go:30]: **nitpick:** typo in `ErrUnkownProperty` -> `ErrUnknownProperty`

**Review by epavlova — APPROVED** [review]

File: ontology/config.yml (hunk 2 above)
Comment (MiroslavGatsanoga) [reply ontology/config.yml:3]: Yeah, I am fine with that, in my initial implementation I did it that way but after that decided to make it more explicit. Will make a note to decide about this when we extract the library as the plan is to review the config structure then as a whole.

File: ontology/config.go (hunk 1 above)
Comment (MiroslavGatsanoga) [reply ontology/config.go:83]: > But what happens if I'm a client of the ontology that uses the method and after everything goes well I decide to do unsafe type assertion to int

Tough luck, I guess 😄 .
After all Go is a strongly typed and this is something to be expected. I think confusion arises here if we user thinks in terms of "I will try to validate the **Go type** of a value with a method that validates its **Ontology type**".
I think we can improve here with being more restrictive about the `int` (Ontology type) to treat as valid `float64` values that are whole integers e.g. "1234.0 is valid, 1234.5 is not valid".

As for adding documentation, I plan to do that once we have stable API for the library which will be established when we extract it as a separate project.

Comment (epavlova) [reply ontology/config.go]: I think that being more restrictive is good idea here as you mentioned - 

> to treat as valid float64 values that are whole integers e.g. "1234.0 is valid, 1234.5 is not valid"

**Review by epavlova — APPROVED** [review]

**Review by ivan-p-nikolov — APPROVED** [review]
LGTM. 

Just something to think about. When extracting the ontology as a lib, we should probably extract `cypher_statements` and `neo_concept` as well in something like `ontology/neo4j` package.

**Pull request merged**